### DISCUSSION TITLE
Keybinding uniformity

### DISCUSSION THREAD

**Discussion by openjck** [helix-editor/helix#10026] [General]
I've been using Vim and Neovim for many years. I'm very interested in Helix, especially because it comes with batteries included, and I've started working through the tutorial. Unfortunately, I have to admit feeling hesitant to continue learning because I dislike some of the default keybindings. I wanted to share my thoughts here. I really don't want to complain, but I would like to start a conversation. It would be interesting to know if others feel similarly, but it would be equally interesting to know if I'm just not getting something.

Some keybindings that I find surprising include:

* Using `map` to select a paragraph, when `m` is otherwise used to **m**atch braces
* Using the space key to perform a wide variety of seemingly-unrelated operations, like opening a file picker, managing windows, and copying to the system clipboard. I recall reading that this might be for compatibility with SpaceVim, but I've never used SpaceVim, and the grouping of space keybindings just seems odd.
* Using `[` and `]` for various kinds of movement, when `g` is also used for movement. I've read this is for compatibility with [vim-unimpaired](https://github.com/tpope/vim-unimpaired), but I've never used vim-unimpaired, and I would have expected these to be `g` keybindings.
* Using `]Space` and `[Space` to insert newlines, when all other `[` and `]` keybindings relate to movement

I don't want to sound unappreciative for the impressive work of the Helix team. On the contrary, I think Helix is really exciting. These are just some things that are preventing me from _falling in love_ with Helix right now.

Does anyone else wish the keybindings were a tad more predictable and uniform? Is there some rhyme or reason for the keybindings that I'm not understanding? Is there any chance these keybindings will change in the future?

Thanks you sharing your thoughts.

**Comment by kirawi** [comment]
The space group is kind of a kitchen sink. I think it's desired to be organized eventually but don't recall what the blocker is.

**Comment by archseer** [comment]
None of the keys are for compatibility, some are simply inspired from popular patterns.

- `m` is match mode, not specific to punctuation. map = match around paragraph. Same for tree sitter objects, e.g. functions etc.
- space is used as a leader key which contains common mappings and a fair bit of LSP functions. We could split some of it out but it's convenient to use the space key since it's easily reachable by the thumb.
- [ and ] are indeed pair mappings similar to vim unimpaired, for example selecting next/previous/first/last error. `]d` is shorter than `gdn`.
- ]Space is a convenience over o<Esc>j and is again a pattern many of us are used to from vim-unimpaired

The good news is all of the keybinds are easily remappable so you can come up with new bindings.] and [ can be hard to type on 40% keyboards so if you come up with a better layout there I'd be interested.
